### PR TITLE
Bug 1255338: Remove scc bound to a project when deleting the project

### PR DESCRIPTION
**Pull request by kargakis**
http://pastebin.com/G7x5hrL6

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1255338

I guess there are things to take into account here:
* A scc can be associated with more than just a service account, we probably don't want to delete it in such a case
* A project can be associated with more than one scc, check all of them and delete only those that don't have any other association?

@pweil- @liggitt

### CODE REVIEW

**Comment by pweil-** [comment]
> A scc can be associated with more than just a service account, we probably don't want to delete it in such a case

We don't want to remove the actual SCC.  An SCC is a cluster wide, reusable object.  We only need to remove any service account associated with the project from the SCC.

> A project can be associated with more than one scc, check all of them and delete only those that don't have any other association

Yes, all SCCs need checked.

I think this probably belongs somewhere in the namespace lifecycle handler for termination.  Examples: https://github.com/pweil-/origin/blob/master/pkg/project/controller/controller.go#L57-57.

This should also occur for a user deletion since SCCs can reference specific users.

**Comment by kargakis** [comment]
> Yes, all SCCs need checked.

Now that I am thinking about it, do we really need to check all SCCs? Since we are going after service acounts and there is one sa per project isn't enough to just delete that sa?

**Comment by kargakis** [comment]
Also if a project is deleted and recreated won't the same happen for the sa?

**Comment by liggitt** [comment]
This is the wrong place to do reference cleanup... we shouldn't start sweeping and cleaning references piece-mill.

Things that can be referenced in a project:
* image streams (from builders or deployments in other projects)
* roles (from policy in other projects)
* service accounts (from policy in other projects, from SCCs)

**Comment by liggitt** [comment]
Also, we should never auto-delete an SCC

**Comment by liggitt** [comment]
> Since we are going after service acounts and there is one sa per project isn't enough to just delete that sa?

A project can have an arbitrary number of service accounts

**Comment by smarterclayton** [comment]
This is garbage collection.  As jordan notes these should be dealt with in that fashion.

File: pkg/project/controller/controller.go [hunk 1]
```diff
@@ -54,48 +54,75 @@ func (c *NamespaceController) Handle(namespace *kapi.Namespace) (err error) {
...
-	err = deleteTemplates(client, namespace)
+	err = deleteTemplates(oc, namespace)
 	if err != nil {
 		return err
 	}
-	return nil
+	return deleteServiceAccounts(kc, namespace)
+}
+
+func deleteServiceAccounts(client kclient.ServiceAccountsNamespacer, ns string) error {
```
Comment (liggitt) [inline pkg/project/controller/controller.go]: the kubernetes namespace controller already cleans up service accounts... we should only be cleaning up origin objects

**Comment by liggitt** [comment]
I think we should table this PR and track the cleanup requirement under a general garbage collection story, which should probably look more like this:

1. Track objects that are referenced (users, groups, roles, service accounts, etc)
2. Periodically sweep referencing objects (rolebindings, sccs, etc), validating and removing stale references

**Comment by kargakis** [comment]
> Track objects that are referenced (users, groups, roles, service accounts, etc)

Now, I see. Track all the references and update the SCCs by removing those.
Users and groups aren't namescoped so we will need to handle them outside of the namespace lifecycle, right?

**Comment by liggitt** [comment]
It doesn't belong in namespace lifecycle... it's a separate operation

**Comment by kargakis** [comment]
K, closing this

**Pull request closed without merging**